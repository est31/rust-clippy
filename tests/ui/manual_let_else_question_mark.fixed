//@run-rustfix
#![allow(unused_braces, unused_variables, dead_code)]
#![allow(
    clippy::collapsible_else_if,
    clippy::unused_unit,
    clippy::let_unit_value,
    clippy::match_single_binding,
    clippy::never_loop
)]
#![warn(clippy::manual_let_else, clippy::question_mark)]

enum Variant {
    A(usize, usize),
    B(usize),
    C,
}

fn g() -> Option<(u8, u8)> {
    None
}

fn e() -> Variant {
    Variant::A(0, 0)
}

fn main() {}

fn foo() -> Option<()> {
    // Fire here, normal case
    let v = g()?;

    // Don't fire here, the pattern is refutable
    let Variant::A(v, w) = e() else { return None };

    // Fire here, the pattern is irrefutable
    let (v, w) = g()?;

    // Don't fire manual_let_else in this instance: question mark can be used instead.
    let v = g()?;

    Some(())
}
