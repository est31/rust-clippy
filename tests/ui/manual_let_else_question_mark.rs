//@run-rustfix
#![allow(unused_braces, unused_variables, dead_code)]
#![allow(
    clippy::collapsible_else_if,
    clippy::unused_unit,
    clippy::let_unit_value,
    clippy::match_single_binding,
    clippy::never_loop
)]
#![warn(clippy::manual_let_else, clippy::question_mark)]

enum Variant {
    A(usize, usize),
    B(usize),
    C,
}

fn g() -> Option<(u8, u8)> {
    None
}

fn e() -> Variant {
    Variant::A(0, 0)
}

fn main() {}

fn foo() -> Option<()> {
    // Fire here, normal case
    let Some(v) = g() else { return None };

    // Don't fire here, the pattern is refutable
    let Variant::A(v, w) = e() else { return None };

    // Fire here, the pattern is irrefutable
    let Some((v, w)) = g() else { return None };

    // Don't fire manual_let_else in this instance: question mark can be used instead.
    let v = if let Some(v_some) = g() { v_some } else { return None };

    Some(())
}
