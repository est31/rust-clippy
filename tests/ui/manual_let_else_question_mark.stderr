error: this `let...else` may be rewritten with the `?` operator
  --> $DIR/manual_let_else_question_mark.rs:30:5
   |
LL |     let Some(v) = g() else { return None };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `let v = g()?;`
   |
   = note: `-D clippy::question-mark` implied by `-D warnings`

error: this `let...else` may be rewritten with the `?` operator
  --> $DIR/manual_let_else_question_mark.rs:36:5
   |
LL |     let Some((v, w)) = g() else { return None };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `let (v, w) = g()?;`

error: this block may be rewritten with the `?` operator
  --> $DIR/manual_let_else_question_mark.rs:39:13
   |
LL |     let v = if let Some(v_some) = g() { v_some } else { return None };
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `g()?`

error: aborting due to 3 previous errors

